### PR TITLE
bridge: Add internal metric to get the current load

### PR DESCRIPTION
**Pull request by martinpitt**
Add a new `sched.loadavg` metric with three instances "1min", "5min", and
"15min", similar to PCP's `kernel.all.load`. This is more efficient than
polling /proc/loadavg and integrates better with the whole metrics
channel flow.

----

This will be used in  https://github.com/martinpitt/performance-graphs/tree/current-metrics

### CODE REVIEW

File: src/bridge/cockpitschedsamples.c [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+  int count;
+
+  count = getloadavg(loadavg, 3);
+  if (count < 0) {
+      g_message ("getloadavg() failed: %m");
+      return;
+    }
+
+  for (int i = 0; i < count; ++i)
+    cockpit_samples_sample (samples, "sched.loadavg", instances[i], (gint64) (loadavg[i] * 100));
```
Comment (martinpitt) [inline src/bridge/cockpitschedsamples.c:39]: Hm, this should get a comment.

Comment (martinpitt) [reply src/bridge/cockpitschedsamples.c:40]: Done.

File: src/bridge/cockpitschedsamples.c [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+void
+cockpit_sched_samples (CockpitSamples *samples)
+{
+  const char * const instances[] = { "1min", "5min", "15min" };
+  double loadavg[3];
+  int count;
+
+  count = getloadavg(loadavg, 3);
+  if (count < 0) {
+      g_message ("getloadavg() failed: %m");
```
Comment (mvollmer) [inline src/bridge/cockpitschedsamples.c:34]: If this error happens at all, I would expect it to happen on every sample, which would spam the journal.  Maybe logging it once is enough?

File: src/bridge/test-metrics.c [hunk 3]
```diff
@@ -719,6 +719,74 @@ test_deprecated_net_all (void)
...
+  cockpit_channel_prepare (channel);
+
+  /* receive meta information */
+  while ((msg = mock_transport_pop_channel (transport, "1234")) == NULL)
+    g_main_context_iteration (NULL, TRUE);
+  res = cockpit_json_parse_bytes (msg, NULL);
+  g_assert (res != NULL);
+  g_assert (json_object_has_member (res, "metrics"));
+
+  /* metrics should have the form [{"name":"sched.loadavg","instances":["1min", "5min", "15min"]}] */
```
Comment (mvollmer) [inline src/bridge/test-metrics.c:752]: Don't we have a `assert_json_equal` kind of function that does the right thing?

Comment (martinpitt) [reply src/bridge/test-metrics.c:752]: I can't use `cockpit_assert_json_eq()` here as the order of the strings is somewhat random. It always *seems* to be "5min,15min,1min", but this looks like some internal glib happenstance which I don't want to pin down in tests.

File: src/bridge/cockpitschedsamples.c (hunk 2 above)
Comment (martinpitt) [reply src/bridge/cockpitschedsamples.c:34]: Good point! Done.

File: src/bridge/test-metrics.c (hunk 3 above)
Comment (allisonkarlitskaya) [reply src/bridge/test-metrics.c:752]: Something about this wasn't sitting right with me so I dug into it deeper.

You might be using the samples interface incorrectly.  The (1,5,15) minute values from the load-average should properly be considered as a tuple.  This puts it on the order of something like the `free`, `used`, `cached`, `swap-used`, etc. values for `memory`.  Those are all handled as separate fields like `memory.free`, `memory.used`, `memory.cached`, etc

All uses of the second parameter that I've found are for iteration over things that are truly generic, such as lists of cgroups in the system, lists of mount points in the system, list of network interfaces, etc.  In that case, the order is indeed arbitrary.

I'd change the way you record samples to instead record them like `loadavg.1min`, etc. with the second parameter always NULL.  Let's discuss :)

**Review by allisonkarlitskaya — CHANGES_REQUESTED** [review]
Cool patch!

I don't like the comment in the commit message about this being more efficient than polling `/proc/loadavg` since this is exactly what we're doing.  Did you mean instead of doing that directly from JS-land?

File: src/bridge/test-metrics.c (hunk 3 above)
Comment (martinpitt) [reply src/bridge/test-metrics.c:752]: I'd like to keep them consistent with PCP's `kernel.all.load` metric, which uses the exact same three instances. That way, JS code can handle a PCP or internal metric channel in the same way, and often code actually ties both as a fallback. [example](https://github.com/cockpit-project/cockpit/blob/master/pkg/dashboard/list.js#L60)_

**Comment by martinpitt** [comment]
> Did you mean instead of doing that directly from JS-land?

Yes, that's what I meant. I updated the commit message accordingly.

File: src/bridge/cockpitschedsamples.c [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+
+void
+cockpit_sched_samples (CockpitSamples *samples)
+{
+  const char * const instances[] = { "1min", "5min", "15min" };
+  double loadavg[3];
+  int count;
+  static bool logged_failure = false;
+
+  count = getloadavg(loadavg, 3);
```
[the remark this answers is not in the document]
Comment (martinpitt) [reply src/bridge/cockpitschedsamples.c:34]: I'd rather not -- man getloadavg(3) says "Up to nelem samples are retrieved and assigned to successive elements of loadavg[].  The system  imposes  a maximum of 3 samples,...", so the code takes care to work with just 1 or 2 samples.

File: src/bridge/cockpitschedsamples.c [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+        {
+          g_message ("getloadavg() failed: %m");
+          logged_failure = true;
+        }
+      return;
+    }
+
+  /* samples are guint64, so multiply by 100 to get two fractional digits */
+  for (int i = 0; i < count; ++i)
+    cockpit_samples_sample (samples, "sched.loadavg", instances[i], (gint64) (loadavg[i] * 100));
```
[the remark this answers is not in the document]
Comment (martinpitt) [reply src/bridge/cockpitschedsamples.c]: Well spotted, thanks! Fixed.

File: src/bridge/test-metrics.c (hunk 3 above)
Comment (mvollmer) [reply src/bridge/test-metrics.c:752]: > I'd like to keep them consistent with PCP's kernel.all.load metric, which uses the exact same three instances. 

Oh, that's bad... I fully agree with @allisonkarlitskaya that the 1 minute load average should be proper metric, and not an instance.

**Comment by martinpitt** [comment]
Converting to draft. I'd like to see what it would take to actually submit doubles instead of this "times 100 to convert to integer" hack -- as that still makes the internal metric incompatible with the pcp one. I also decided to not block performance-graphs on that -- let's just poll /proc/loadavg for the time being, so that this also works on RHEL 8.3.

**Comment by martinpitt** [comment]
Reading /proc/loadavg from JS isn't too bad, and this turned out to be more problematic than I initially thought. Closing for now, we can dig this out again if/when we run into a case where this is worth the effort.

**Pull request closed without merging**